File: mcp-servers/packages/core/src/graphClient.ts
import type { TokenCredential } from "@azure/identity";

export const GRAPH_BASE = "https://graph.microsoft.com/v1.0";
export const GRAPH_DEFAULT_SCOPE = "https://graph.microsoft.com/.default";

/** Bounded retry budget for 429 (throttling). Graph throttles routinely in real
 * tenants -- dying on the first 429 makes a tool unusable, but retrying forever
 * risks hanging a caller, so this caps both attempts and the wait per attempt. */
const MAX_RETRY_ATTEMPTS = 3;
const MAX_RETRY_WAIT_MS = 30_000;

export class GraphApiError extends Error {
  status: number;
  body: unknown;

  constructor(status: number, body: unknown, url: string) {
    super(`Graph API request to ${url} failed with status ${status}: ${describeBody(body)}`);
    this.name = "GraphApiError";
    this.status = status;
    this.body = body;
  }
}

function describeBody(body: unknown): string {
  if (typeof body === "string") return body.slice(0, 500);
  try {
    return JSON.stringify(body);
  } catch {
    return String(body);
  }
}

/**
 * Reads a response body without assuming it's JSON. Graph errors are normally
 * JSON, but a WAF, a proxy, or a plain 5xx from an intermediate hop can return
 * HTML or plain text -- JSON.parse on that throws SyntaxError and loses the
 * real status code inside a confusing secondary error. Falls back to the raw
 * text (truncated by describeBody above when it ends up in an error message).
 */
async function readBody(res: Response): Promise<unknown> {
  if (res.status === 204) return undefined;
  const text = await res.text();
  if (!text) return undefined;
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

function retryDelayMs(res: Response, attempt: number): number {
  const header = res.headers.get("retry-after");
  let ms = 1000 * 2 ** attempt; // exponential backoff fallback if the header is absent/unusable
  if (header) {
    const seconds = Number(header);
    if (Number.isFinite(seconds)) {
      ms = seconds * 1000;
    } else {
      const at = Date.parse(header);
      if (!Number.isNaN(at)) ms = at - Date.now();
    }
  }
  if (!Number.isFinite(ms) || ms < 0) ms = 1000 * 2 ** attempt;
  return Math.min(ms, MAX_RETRY_WAIT_MS);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export interface GraphRequestOptions {
  query?: Record<string, string | number | undefined>;
  body?: unknown;
  /** Extra scopes for the token request. Defaults to the Graph .default scope. */
  scopes?: string[];
}

export interface GraphPage<T> {
  items: T[];
  /** True when the page cap (maxPages) was hit while @odata.nextLink still had
   * more to fetch -- the caller has less than the full result set and should
   * say so rather than presenting it as complete. */
  truncated: boolean;
}

/**
 * Minimal typed wrapper over fetch() for Microsoft Graph. Deliberately does
 * not pull in the full @microsoft/microsoft-graph-client SDK -- these
 * servers only ever touch a couple dozen endpoints each, and a thin client
 * is easier to mock offline in tests (see test/graphClient.test.ts).
 */
export class GraphClient {
  constructor(
    private readonly credential: TokenCredential,
    private readonly baseUrl: string = GRAPH_BASE,
    /** Scopes used when a call doesn't pass its own. App-only clients keep the
     * .default scope; delegated (device code) clients pass the specific
     * Graph permissions they need at construction time. */
    private readonly defaultScopes: string[] = [GRAPH_DEFAULT_SCOPE]
  ) {}

  private async getToken(scopes: string[]): Promise<string> {
    const token = await this.credential.getToken(scopes);
    if (!token) {
      throw new Error("Failed to acquire a Graph access token (credential.getToken returned null).");
    }
    return token.token;
  }

  private buildUrl(path: string, query?: GraphRequestOptions["query"]): string {
    const url = path.startsWith("http") ? new URL(path) : new URL(this.baseUrl + path);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined) url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }

  /** fetch() with a bounded 429 retry, honoring Retry-After when Graph sends one. */
  private async fetchWithRetry(url: string, init: RequestInit): Promise<Response> {
    let attempt = 0;
    for (;;) {
      const res = await fetch(url, init);
      if (res.status === 429 && attempt < MAX_RETRY_ATTEMPTS) {
        await sleep(retryDelayMs(res, attempt));
        attempt += 1;
        continue;
      }
      return res;
    }
  }

  async request<T = unknown>(
    method: "GET" | "POST" | "PATCH" | "DELETE" | "PUT",
    path: string,
    options: GraphRequestOptions = {}
  ): Promise<T> {
    const scopes = options.scopes ?? this.defaultScopes;
    const token = await this.getToken(scopes);
    const url = this.buildUrl(path, options.query);

    const res = await this.fetchWithRetry(url, {
      method,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
    });

    const parsed = await readBody(res);

    if (!res.ok) {
      throw new GraphApiError(res.status, parsed, url);
    }

    return parsed as T;
  }

  get<T = unknown>(path: string, options?: GraphRequestOptions): Promise<T> {
    return this.request<T>("GET", path, options);
  }
  /** WRITE. Callers must route this through core's assertWriteAllowed (env flag
   * + confirm:true) before calling -- this method performs no gating itself. */
  post<T = unknown>(path: string, options?: GraphRequestOptions): Promise<T> {
    return this.request<T>("POST", path, options);
  }
  /** WRITE. Callers must route this through core's assertWriteAllowed (env flag
   * + confirm:true) before calling -- this method performs no gating itself. */
  patch<T = unknown>(path: string, options?: GraphRequestOptions): Promise<T> {
    return this.request<T>("PATCH", path, options);
  }
  /** WRITE. Callers must route this through core's assertWriteAllowed (env flag
   * + confirm:true) before calling -- this method performs no gating itself. */
  put<T = unknown>(path: string, options?: GraphRequestOptions): Promise<T> {
    return this.request<T>("PUT", path, options);
  }
  /** WRITE/DESTRUCTIVE. Callers must route this through core's assertWriteAllowed
   * (env flag + confirm:true) before calling -- this method performs no gating
   * itself. */
  delete<T = unknown>(path: string, options?: GraphRequestOptions): Promise<T> {
    return this.request<T>("DELETE", path, options);
  }

  /** Follows @odata.nextLink until exhausted or maxPages is reached. Reports
   * truncation rather than silently returning a partial result as if it were
   * complete -- see GraphPage. */
  async getAllPages<T = unknown>(
    path: string,
    options: GraphRequestOptions & { maxPages?: number } = {}
  ): Promise<GraphPage<T>> {
    const maxPages = options.maxPages ?? 10;
    const items: T[] = [];
    let next: string | undefined = this.buildUrl(path, options.query);
    let page = 0;

    while (next && page < maxPages) {
      const scopes = options.scopes ?? this.defaultScopes;
      const token = await this.getToken(scopes);
      const res = await this.fetchWithRetry(next, {
        headers: { Authorization: `Bearer ${token}` },
      });
      const parsed = await readBody(res);
      if (!res.ok) {
        throw new GraphApiError(res.status, parsed, next);
      }
      const body = (parsed ?? {}) as { value?: T[]; "@odata.nextLink"?: string };
      items.push(...(body.value ?? []));
      next = body["@odata.nextLink"];
      page += 1;
    }

    return { items, truncated: Boolean(next) && page >= maxPages };
  }
}

File: mcp-servers/core/src/mcp_ms_core/graph_client.py
"""A small, paging- and throttling-aware Microsoft Graph HTTP client.

Graph throttles aggressively (429) and buries its real error message inside
the response body as JSON-within-JSON - this wraps both so every server's
tool code just calls ``get``/``post``/``patch``/``delete``/``get_all`` and
either gets data back or a ``GraphError`` with an actionable message.
"""

from __future__ import annotations

import time
from typing import Any, Callable

import httpx

from .errors import GraphError

GRAPH_V1 = "https://graph.microsoft.com/v1.0"
GRAPH_BETA = "https://graph.microsoft.com/beta"

_MAX_ATTEMPTS = 4
_RETRY_STATUS = {429, 500, 502, 503, 504}


class GraphClient:
    """Wraps an ``httpx.Client``. Pass a zero-arg ``token_provider`` callable -
    normally ``ClientCredentialAuth.get_token`` or ``DeviceCodeAuth.get_token``
    - so the client always sends a fresh (or silently-refreshed) bearer token
    rather than caching one that can expire mid-session.
    """

    def __init__(
        self,
        token_provider: Callable[[], str],
        *,
        base_url: str = GRAPH_V1,
        timeout: float = 30.0,
        transport: httpx.BaseTransport | None = None,
    ) -> None:
        self._token_provider = token_provider
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(timeout=timeout, transport=transport)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "GraphClient":
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self._token_provider()}",
            "Accept": "application/json",
        }

    def _url(self, path: str) -> str:
        if path.startswith("https://"):
            return path
        return f"{self._base_url}/{path.lstrip('/')}"

    def request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json_body: Any = None,
    ) -> dict[str, Any]:
        url = self._url(path)
        resp: httpx.Response | None = None
        for attempt in range(_MAX_ATTEMPTS):
            resp = self._client.request(
                method, url, params=params, json=json_body, headers=self._headers()
            )
            if resp.status_code in _RETRY_STATUS and attempt < _MAX_ATTEMPTS - 1:
                delay = _retry_delay(resp, attempt)
                time.sleep(delay)
                continue
            break
        assert resp is not None
        if resp.status_code >= 400:
            raise _to_graph_error(resp)
        if resp.status_code == 204 or not resp.content:
            return {}
        return resp.json()

    def get(self, path: str, *, params: dict[str, Any] | None = None) -> dict[str, Any]:
        return self.request("GET", path, params=params)

    def post(self, path: str, *, json_body: Any = None) -> dict[str, Any]:
        return self.request("POST", path, json_body=json_body)

    def patch(self, path: str, *, json_body: Any = None) -> dict[str, Any]:
        return self.request("PATCH", path, json_body=json_body)

    def delete(self, path: str) -> dict[str, Any]:
        return self.request("DELETE", path)

    def get_all(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        limit: int | None = None,
    ) -> list[dict[str, Any]]:
        """Follow ``@odata.nextLink`` until exhausted or ``limit`` items collected.

        For a fleet-wide pull of more than a few hundred records, prefer a
        dedicated export/report endpoint where the API offers one - looping
        this over tens of thousands of objects is slow and throttles hard.
        """
        items: list[dict[str, Any]] = []
        next_path: str | None = path
        next_params: dict[str, Any] | None = params
        while next_path:
            page = self.request("GET", next_path, params=next_params)
            items.extend(page.get("value", []))
            if limit is not None and len(items) >= limit:
                return items[:limit]
            next_path = page.get("@odata.nextLink")
            next_params = None  # nextLink is already a full URL with params baked in
        return items


def _retry_delay(resp: httpx.Response, attempt: int) -> float:
    retry_after = resp.headers.get("Retry-After")
    if retry_after is not None:
        try:
            return min(float(retry_after), 30.0)
        except ValueError:
            pass
    return min(2.0**attempt, 30.0)


def _to_graph_error(resp: httpx.Response) -> GraphError:
    request_id = resp.headers.get("request-id") or resp.headers.get("client-request-id")
    code = "unknown"
    message = resp.text[:300]
    try:
        body = resp.json()
        err = body.get("error", {})
        code = err.get("code", code)
        message = err.get("message", message)
    except ValueError:
        pass
    return GraphError(resp.status_code, code, message, request_id=request_id)
